### ISSUE TITLE
Join table with extra columns

### ISSUE THREAD

**Issue by manubb** [Vincit/objection.js#97]
Hello.
We are currently using join tables with extra columns and we had to use some workarounds with objection and objection-rest. I would like to suggest several changes to objection.js to improve support of such join tables. For example, let say that persons have ordered movie collections. It would be nice to be able to write something like:
```javascript
movies: {
 relation: Model.ManyToManyRelation,
 modelClass: __dirname + '/Movie',
 join: {
  from: 'Person.id',
  through: {
   from: 'Person_Movie.personId',
   to: 'Person_Movie.movieId',
   extra: ['order']
  },
  to: 'Movie.id'
 }
}
```
so that:
```javascript
person.$relatedQuery('movies')
```
 automatically retrieves the order of each movie. It would be also be great to have something like:
```javascript
person.$relatedQuery('movies').relate({id: 50, order: 5});
person.$relatedQuery('movies').unrelate().where('id', 50).where('order', 3);`
```
What do you think about this?
Manuel

**Comment by koskimas** [comment]
Sorry for the delay.

Good idea! This is something that has been missing. Only thing I would change is the name of the `extra` property. Could we come up with something more descriptive? `extraProperties`? May be too long and still not descriptive. `props`? `properties`? This is the hardest part of programming :smile: 

Naturally this would also work in these cases:

```js
Person
  .query()
  .insertWithRelated({
    movies: [{
      name: 'The room',
      order: 1
    }]
  })
```

```js
Movie
  .query()
  .insert({
    name: 'The room',
    order: 1
  })
```

**Comment by kapouer** [comment]
Shouldn't there be a ModelClass associated with "through" instead of a list of columns ?

**Comment by koskimas** [comment]
You can already accociate a model for a join table like this:

```js
through: {
  from: 'foo.bar',
  to: 'baz.spam',
  modelClass: JoinModel
}
```

but where would you get the columns to join? From jsonSchema? I'd like to avoid making jsonSchema a mandatory part of the models since many people don't like to use it. It can even be argued that validation is not the job for an ORM.

**Comment by Hnus-** [comment]
Support for this would eliminate some very ugly code on my back-end I am looking forward to this future.

**Comment by koskimas** [comment]
@kapouer I get your point though. Do you have any suggestion how this could be handled using the model class?

@Hnus- as you know,`through` describes the join table. In my mind `related` would imply that the given properties are somehow related to the join table but not the join table's own properties.

How about `include` or `includeProperties`?

**Comment by kapouer** [comment]
yes, *if* one sets `join.through.modelClass` then `person.$relatedQuery('movies')` /should/ fetch all known properties from the join table (except the ones used in `from` and `to`) by default ?
Maybe what is needed here is more how to deal with column name conflicts ?
In the example given by @manubb 
```
person.$relatedQuery('movies').relate({id: 50, "person_movie.order": 5});
```
i'm sure this idea of extra column could be a little more generic and coherent with objection.js style.

**Comment by koskimas** [comment]
Consider that same example relation from the other direction: `Movie` has a list of `Person`s as `actors` relation. You would want to use the same model class for that relation also, but automatically joining the `order` for each actor would make little sense. I think we need to be able to control the joined properties per relation an not per join table. There are other cases also. I have written an N-way join table (where N > 2) many times, and it would't automatically make sense to join all the properties of the join table through each relation that uses it.

The reason I often create a model class for the join table is that I can create a REST API for the join model. This way you can achieve the same thing we are trying to solve here. For example `relate` is simply an insert to the join model. Insert can be achieved with a `insertWithRelated` operation to the join model and so on. For example:

```js
// Relate a person 123 to a movie 567
PersonMovie
  .query()
  .insert({
    order: 1,
    personId: 123,
    movieId: 567
  });
```

```js
// Insert a new movie and relate it to person 123
PersonMovie
  .query()
  .insertWithRelated({
    order: 1,
    personId: 123
    movie: {
      name: 'The room'
    }
  });
```

```js
// Insert a new person and relate it to movie 567
PersonMovie
  .query()
  .insertWithRelated({
    order: 35,
    movieId: 567
    person: {
      firstName: 'Tommy',
      lastName: 'Wiseau'
    }
  });
```

**Comment by koskimas** [comment]
Implemented in 0.5.0-rc.1. Run `npm install objection@next` to install. The property name in `through` object is the originally suggested `extra`.

**Comment by ilkka-rautiainen** [comment]
A great feature, but I found a weakness in it: there's no means of coping with colliding column names between the join table and the target table. Now the join table values simply override the values in target table. The clearest solution might be to separate the join-object and the target-object somehow.
